### PR TITLE
MI-2347: Add Keyboard accessibility in Select and Autocomplete component

### PR DESCRIPTION
**Pull request by tanisha-agarwal**
### Ticket Link
[MI-2347](https://brightscout.atlassian.net/browse/MI-2347)

### Description
Add Keyboard accessibility in the Select and Autocomplete component

### CODE REVIEW

File: src/components/AutoComplete/AutoComplete.component.tsx [hunk 1]
```diff
@@ -136,6 +139,30 @@ export const AutoComplete = (props: AutoCompleteProps) => {
...
+			setSearchValue(option.label ?? option.value);
+			setActive(0);
+			setOptions([]);
+			setOpen(false);
+    }
+		else if (event.key === 'ArrowUp') {
+      if (active === 0) return;
+			setActive(prev => prev - 1);
+			listRef.current.scrollBy(0, -Constants.ITEM_HEIGHT);
+    }
```
Comment (ayusht2810) [inline src/components/AutoComplete/AutoComplete.component.tsx:157]: fix indentations everywhere

File: src/components/Select/Select.component.tsx [hunk 2]
```diff
@@ -132,17 +162,18 @@ export const Select = (props: SelectProps) => {
 		<Wrapper>
 			<Input
 				ref={inputRef}
-				placeholder=''
+				placeholder=' '
```
Comment (ayusht2810) [inline src/components/Select/Select.component.tsx:165]: why this change?

File: src/components/Select/Select.component.tsx [hunk 3]
```diff
@@ -81,23 +85,47 @@ export const Select = (props: SelectProps) => {
...
+		if (event.key === 'Enter') {
+			const option = options[active];
+			setValue(option.label ?? option.value);
+			onSelectOptionHandler(event, option);
+			setIsOpen(false);
+		}
+		else if (event.key === 'ArrowUp') {
+			if (active === 0) return;
+			setActive((prev) => prev - 1);
+			listRef.current.scrollBy(0, -37);
```
Comment (ayusht2810) [inline src/components/Select/Select.component.tsx:121]: use constant `ITEM_HEIGHT` here

File: src/components/Select/Select.component.tsx [hunk 4]
```diff
@@ -81,23 +85,47 @@ export const Select = (props: SelectProps) => {
...
+		}
+		else if (event.key === 'ArrowDown') {
+			if (active === options.length - 1) return;
+			setActive((prev) => prev + 1);
+			listRef.current.scrollBy(0, 37);
```
Comment (ayusht2810) [inline src/components/Select/Select.component.tsx:126]: does this `listRef` also reset on closing the dropdown?

File: src/components/Select/Select.component.tsx [hunk 5]
```diff
@@ -81,23 +85,47 @@ export const Select = (props: SelectProps) => {
...
+
+	/**
+	 * The function is called when an event is detected on the keyboard,
+	 * so you can browse through the list and select one.
+	 */
+	const onKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (event) => {
+		if (event.key === 'Enter') {
+			const option = options[active];
+			setValue(option.label ?? option.value);
+			onSelectOptionHandler(event, option);
```
Comment (ayusht2810) [inline src/components/Select/Select.component.tsx:115]: don't we need to set `setActive(0)` here?

File: src/components/Select/Select.component.tsx (hunk 2 above)
Comment (tanisha-agarwal) [reply src/components/Select/Select.component.tsx:165]: In the select component, we are checking if the placeholder is empty of not
if the placeholder is empty, then the label will align b/w border else inside input

File: src/components/Select/Select.component.tsx (hunk 5 above)
Comment (tanisha-agarwal) [reply src/components/Select/Select.component.tsx:115]: I think no, because `on enter`, the active item will be the selected one

**Review by ayusht2810 — APPROVED** [review]

File: src/components/AutoComplete/AutoComplete.component.tsx [hunk 6]
```diff
@@ -107,6 +110,17 @@ export const AutoComplete = (props: AutoCompleteProps) => {
 		};
 	}, []);
 
+	/**
+	 * If isOpen is true and value is empty, then set the active index to 0 and scroll the list to (0,0)
```
Comment (avas27JTG) [inline src/components/AutoComplete/AutoComplete.component.tsx:114]:
```suggestion
	 * If "isOpen" is true and value is empty, then set the active index to 0 and scroll the list to (0,0)
```
we can treat variable name as noun and wrap them in `" "` to signify that

File: src/components/AutoComplete/AutoComplete.component.tsx [hunk 7]
```diff
@@ -135,6 +149,30 @@ export const AutoComplete = (props: AutoCompleteProps) => {
...
+			setSearchValue(option.label ?? option.value);
+			setActive(0);
+			setOptions([]);
+			setOpen(false);
+    }
+		else if (event.key === 'ArrowUp') {
+      if (active === 0) return;
+			setActive(prev => prev - 1);
+			listRef.current.scrollBy(0, -Constants.ITEM_HEIGHT);
+    }
```
Comment (avas27JTG) [inline src/components/AutoComplete/AutoComplete.component.tsx:168]: fix indentations

File: src/components/AutoComplete/AutoComplete.component.tsx [hunk 8]
```diff
@@ -135,6 +149,30 @@ export const AutoComplete = (props: AutoCompleteProps) => {
...
+		else if (event.key === 'ArrowDown') {
+      if (active === options.length - 1) return;
+      setActive(prev => prev + 1);
+			listRef.current.scrollBy(0, Constants.ITEM_HEIGHT);
+    }
```
Comment (avas27JTG) [inline src/components/AutoComplete/AutoComplete.component.tsx:173]:
```suggestion
		if (event.key === 'Enter') {
			const option = options[active];
			setSearchValue(option.label ?? option.value);
			setActive(0);
			setOptions([]);
			setOpen(false);
			return;
    }
    
	if (event.key === 'ArrowUp') {
            if (active === 0) return;
			setActive(prev => prev - 1);
			listRef.current.scrollBy(0, -Constants.ITEM_HEIGHT);
			return;
    }
    
    if (event.key === 'ArrowDown') {
            if (active === options.length - 1) return;
            setActive(prev => prev + 1);
			listRef.current.scrollBy(0, Constants.ITEM_HEIGHT);
			return;
    }
```

we should prevent `else if` by using `return` if possible this makes code more readable and prevents dependency flow

File: src/components/AutoComplete/AutoComplete.component.tsx [hunk 9]
```diff
@@ -158,17 +197,20 @@ export const AutoComplete = (props: AutoCompleteProps) => {
 				}}
 				{...restProps}
 			/>
-			{options.length && (
+			{!!options.length && (
```
Comment (avas27JTG) [inline src/components/AutoComplete/AutoComplete.component.tsx:200]: `!!` is doing the same thing as `!!false = !true = false`, why are we doing this?

File: src/components/Select/Select.component.tsx [hunk 10]
```diff
@@ -81,40 +87,68 @@ export const Select = (props: SelectProps) => {
...
+			setIsOpen(false);
+		} else if (event.key === 'ArrowUp') {
+			if (active === 0) return;
+			setActive((prev) => prev - 1);
+			listRef.current.scrollBy(0, -Constants.ITEM_HEIGHT);
+		} else if (event.key === 'ArrowDown') {
+			if (active === options.length - 1) return;
+			setActive((prev) => prev + 1);
+			listRef.current.scrollBy(0, Constants.ITEM_HEIGHT);
+		}
```
Comment (avas27JTG) [inline src/components/Select/Select.component.tsx:130]: same as above use `return`

File: src/constants/index.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export const Constants = {
+    ITEM_HEIGHT: 37,
+    SUGGESTION_FETCH_DELAY: 500,
```
Comment (avas27JTG) [inline src/constants/index.ts:3]:
```suggestion
    FETCH_SUGGESTION_DELAY: 500,
```

File: src/components/AutoComplete/AutoComplete.component.tsx (hunk 9 above)
Comment (tanisha-agarwal) [reply src/components/AutoComplete/AutoComplete.component.tsx:200]: when using `{options.length && ` if the options array was empty it was showing 0 on UI. That's why have to convert `options.length` condition to boolean either by `options.length > 0` or `!!options.length`

Comment (avas27JTG) [reply src/components/AutoComplete/AutoComplete.component.tsx:205]: see if this works `!Boolean(options.length)`

**Review by avas27JTG — APPROVED** [review]
